### PR TITLE
Allow using OIDC to fetch api tokens

### PR DESCRIPTION
**Pull request by segiddins**
Implements https://github.com/rubygems/rfcs/pull/49

### CODE REVIEW

**Review by indirect — APPROVED** [review]
Are there docs somewhere for how to use this, either as a maintainer or as a developer publishing gems? I'm ok merging this without, but I do think we should have docs to point people at when we ask them to help us test it!

**Comment by segiddins** [comment]
@indirect right now, we have the [RFC](https://github.com/rubygems/rfcs/pull/49) with a "guide level explanation" and the README to the [recommended GH action](https://github.com/rubygems/configure-rubygems-credentials).

I'd like to wait to migrate the contents to the guides proper until we're ready to properly "release" the feature

**Comment by indirect** [comment]
Sounds good.

File: app/controllers/api/v1/oidc/api_key_roles_controller.rb
Comment (duckinator) [inline app/controllers/api/v1/oidc/api_key_roles_controller.rb:80]: It might make sense to open a separate issue about this if you don't resolve it in the PR itself. (Just so it's in a more visible location.)

**Review by duckinator — COMMENTED** [review]
I'm going to leave the "official" reviews to folks more familiar with the codebase, but figured an extra pair of eyes couldn't hurt given how big this PR is. I don't see any issues, and I'm excited that progress is being made on this! 🙂

**Comment by segiddins** [comment]
Heads up, I will merge this on Friday unless any blocking reviews are left

File: app/controllers/passwords_controller.rb
Comment (jenshenny) [inline app/controllers/passwords_controller.rb:26]: why not update the `soft_deleted_at` vs `expires_at`? what is the difference between the two (`soft_deleted_at` is used when the user doesn't own the gem that is scoped to a key IIRC) . Could we consolidate them into one concept/attribute since they both convey that the api key cannot be used anymore.

**Review by jenshenny — COMMENTED** [review]
Looks good! Put down a few initial questions.

File: app/controllers/api/v1/oidc/api_key_roles_controller.rb
Comment (segiddins) [reply app/controllers/api/v1/oidc/api_key_roles_controller.rb:30]: im a bit confused, it's surfaced right here in the API?

File: app/controllers/api_keys_controller.rb
Comment (segiddins) [reply app/controllers/api_keys_controller.rb:10]: they work identically to each other, the only difference is how they are created

for now, im avoiding touching the HTML pages, wanting to wait until we're satisfied with the bones of this and ready to roll it out to GA to do that work

File: app/controllers/passwords_controller.rb
Comment (segiddins) [reply app/controllers/passwords_controller.rb:26]: because soft_deleted_at is specifically associated with the associated gem no longer being owned by the user, and I don't want to change that meaning.

Also, right now we only check for presence of soft_deleted_at, and not whether its in the past or future

File: app/controllers/api_keys_controller.rb
Comment (jenshenny) [reply app/controllers/api_keys_controller.rb:10]: Ah I see that the key name is namespaced by the name of the API key role so a user can differentiate it by that. I would want to see these keys under a [API key role section](https://github.com/rubygems/rubygems.org/pull/3716/files#r1284002994) at some point where I can see all keys up to the last x days (user can only see keys that are active and can't see soft deleted keys - even better show the [api key role to a specific gem push](https://github.com/rubygems/rfcs/blob/master/text/0010-OIDC.md?plain=1#L203)) 

Curious if this is what you had in mind.


Eg.
```
#API Key roles

##<name>
Scopes:
Expiry:
Gems:

<list of keys>
```

**Review by jenshenny — COMMENTED** [review]
Was able to test this out! Super cool to see this in action ✨ Thanks for putting in all this work

File: app/controllers/api_keys_controller.rb
Comment (segiddins) [reply app/controllers/api_keys_controller.rb:10]: honestly, I haven't thought about the HTML yet, my plan is to start on that once this ships

File: app/controllers/api/v1/oidc/api_key_roles_controller.rb
Comment (jenshenny) [reply app/controllers/api/v1/oidc/api_key_roles_controller.rb:30]: Sorry it was more related to surfacing this info in the html and redundant to this thread https://github.com/rubygems/rubygems.org/pull/3716#discussion_r1284007980

File: app/avo/resources/oidc_provider_resource.rb
Comment (jenshenny) [reply app/avo/resources/oidc_provider_resource.rb:15]: Agreed it's not blocking. Just raising that this can be iterated on so other admins are able to create/modify providers more easily when it comes to that point.

File: app/controllers/api_keys_controller.rb
Comment (jenshenny) [reply app/controllers/api_keys_controller.rb:10]: Gotcha, sounds great! I think this is necessary enough to be included in the first release and not for GA.

Comment (segiddins) [reply app/controllers/api_keys_controller.rb:10]: and here it is :)

File: app/avo/resources/oidc_provider_resource.rb
Comment (segiddins) [reply app/avo/resources/oidc_provider_resource.rb:15]: The modification is done automatically via an action that fetches from the well-known URLs

File: config/locales/en.yml
Comment (jenshenny) [reply config/locales/en.yml:724]: Would users understand what an API key **role** actually is? It reveals too much of the implementation details. I would prefer `provider` or `issuer` even though it does conflict with some fields in the model.

```suggestion
        api_key_roles: API key provider (OIDC) 
      show:
        api_key_role_name: "API key provider %{name}"
```

File: app/views/oidc/api_key_roles/index.html.erb
Comment (jenshenny) [inline app/views/oidc/api_key_roles/index.html.erb:28]: When this is setup on prod, could we name the issuer `Github Actions`? is `provider.issuer` the same as `provider.configuration.issuer`? If so, could we remove `provider.issuer` and add a name attribute and store it as `Github Actions`.

Screenshot 2023-08-14 at 8 16 58 PM

File: app/views/oidc/api_key_roles/index.html.erb
Comment (jenshenny) [inline app/views/oidc/api_key_roles/index.html.erb:22]: It's a bit hard to tell that the link goes to the show page. Can we add a `Details` button on the side of each row like the edit button in API keys?
Screenshot 2023-08-14 at 8 25 11 PM

File: app/views/oidc/api_key_roles/show.html.erb
Comment (jenshenny) [inline app/views/oidc/api_key_roles/show.html.erb:17]: Showing the raw json is a bit hard to read. Why not use the scopes, valid_for and gems from `OIDC::ApiKeyPermissions`?

File: app/views/api_keys/index.html.erb
Comment (jenshenny) [inline app/views/api_keys/index.html.erb:112]: nit: I would prefer to have the api_key_roles index view on the same page rather than having it separate. The button placement looks a bit off.

File: app/views/oidc/api_key_roles/show.html.erb
Comment (jenshenny) [inline app/views/oidc/api_key_roles/show.html.erb:45]: This is pretty overwhelming. I think we can just get by with expired_at, created_at, and jti.

File: app/views/oidc/api_key_roles/show.html.erb
Comment (jenshenny) [inline app/views/oidc/api_key_roles/show.html.erb:32]: Instead of Jti for this, could we set it to `JWT ID`?

File: app/views/oidc/api_key_roles/show.html.erb
Comment (jenshenny) [inline app/views/oidc/api_key_roles/show.html.erb:21]: This is fine for now, but for GA, I don't want to show raw json.

**Review by jenshenny — COMMENTED** [review]
I have some comments regarding the UX of the basic views.

The PR is getting kind of large though. I would be 👍 in merging everything except https://github.com/rubygems/rubygems.org/pull/3716/commits/73bf7c9c67380889da19c72b55cda886d6cc5a35 and have another PR for the views. I would also be open to making the enhancements to the views too! Just let me know.

**Comment by segiddins** [comment]
I mean, i didnt want to do the HTML in this PR but it seemed like it was a blocker

**Comment by jenshenny** [comment]
Apologies if it sounded like the HTML was a blocker for this PR specifically 😅  To rephrase my [original comment](https://github.com/rubygems/rubygems.org/pull/3716#discussion_r1284911204), I agree with your plan to work on it after this PR was merged, as long as the views are going to be created before any announcements of the alpha release is going to be sent out.

**Comment by segiddins** [comment]
@jenshenny reverted the HTML, will revisit it as soon as this merges (will keep going on the same branch so we can use the same shipit env for it).

I don't plan on announcing the alpha -- the goal is to get a handful of folks (e.g. those on the rubygems / rubygems.org teams) using the feature before announcing it at all / opening it up for wider adoption. That way, we can make any adjustments as needed before committing to the API in any way.

**Review by jenshenny — APPROVED** [review]
> I don't plan on announcing the alpha -- the goal is to get a handful of folks (e.g. those on the rubygems / rubygems.org teams) using the feature before announcing it at all / opening it up for wider adoption. That way, we can make any adjustments as needed before committing to the API in any way.

Oh ok, I misunderstood what alpha entails - that makes sense, thanks for clarifying!

**Pull request merged**
